sql: extend FROM[...], introduce SHOW SESSION TRACE/SHOW TRACE FOR

knz opened the pull request:
### sql: generalize the FROM [...] syntax and PREPARE

This patch extends the statement-as-data-source syntax to support any
statement that can also be EXPLAINed. This removes the need for
multiple separate grammar rules, and introduces support for things
like `SELECT ... FROM [INSERT ... RETURNING ...] WHERE ...`,
or more crazy like
`INSERT INTO ... SELECT * FROM [INSERT ... RETURNING ...]`

### sql: introduce SHOW SESSION TRACE, replace EXPLAIN(TRACE) by SHOW TRACE FOR

Two changes in this patch:

- the new statement `SHOW SESSION TRACE` is introduced, as the
  user-facing interface to `TABLE crdb_internal.session_trace`.  The
  alias is introduced because we want to avoid documenting the
  contents of `crdb_internal` too much, whereas session tracing is a
  generally useful feature.

- UX improvement: previously `EXPLAIN(TRACE)` was able to operate on
  any "explainable statement", which includes data-modifying
  statements like INSERT or ALTER TABLE. This was a UX
  surprise/problem because on the one hand, intuition says that
  EXPLAIN should merely describe the behavior but not "do" anything
  (as in, not modify data) whereas on the other hand the
  implementation *is* running the statement and data is modified.

  This patch solves the issue by moving the functionality of
  `EXPLAIN(TRACE)` under **a new statement `SHOW TRACE FOR`**.

  Once `EXPLAIN(DEBUG)` (which suffers from the same problem) is also
  removed in a subsequent patch, users will gain confidence when they
  execute EXPLAIN that no data is modified.

  For now `SHOW TRACE FOR` is using the same underlying mechanism as
  `EXPLAIN(TRACE)` was using previously (`explainTraceNode`). A
  subsequent patch can simplify that to use the same session tracing
  as `SHOW SESSION TRACE`.

### sql: simplify SHOW TRACE FOR to use session tracing

This replaces the underlying mechanism to use session tracing. This removes the `hijackCtx` helpers which were an eyesore.

Needed for #16514.

knz committed:
sql: introduce SHOW SESSION TRACE, replace EXPLAIN(TRACE) by SHOW TRACE FOR

Two changes in this patch:

- the new statement `SHOW SESSION TRACE` is introduced, as the
  user-facing interface to `TABLE crdb_internal.session_trace`.  The
  alias is introduced because we want to avoid documenting the
  contents of `crdb_internal` too much, whereas session tracing is a
  generally useful feature.

- UX improvement: previously `EXPLAIN(TRACE)` was able to operate on
  any "explainable statement", which includes data-modifying
  statements like INSERT or ALTER TABLE. This was a UX
  surprise/problem because on the one hand, intuition says that
  EXPLAIN should merely describe the behavior but not "do" anything
  (as in, not modify data) whereas on the other hand the
  implementation *is* running the statement and data is modified.

  This patch solves the issue by moving the functionality of
  `EXPLAIN(TRACE)` under **a new statement `SHOW TRACE FOR`**.

  Once `EXPLAIN(DEBUG)` (which suffers from the same problem) is also
  removed in a subsequent patch, users will gain confidence when they
  execute EXPLAIN that no data is modified.

  For now `SHOW TRACE FOR` is using the same underlying mechanism as
  `EXPLAIN(TRACE)` was using previously (`explainTraceNode`). A
  subsequent patch can simplify that to use the same session tracing
  as `SHOW SESSION TRACE`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/16524)
<!-- Reviewable:end -->

RaduBerinde commented:
What's the reason for only supporting selects?

knz commented:
The main reason is in the commit message (intuition given by EXPLAIN)

Another strong technical reason is that our statement parallelizer is unable to dig "into" `SELECT * FROM [EXPLAIN INSERT ...]` to see there is an INSERT hidden in there.

RaduBerinde commented:
Well I am confused. You explain that the intuition of EXPLAIN makes it surprising that an EXPLAIN statement mutates data. I agree. But you are renaming it to `SHOW TRACE FOR`. I don't think it's still surprising with the new syntax.

knz commented:
Hmm it seems we are both confused.

I think I simply missed another important argument in the commit message. will add

knz commented:
I have extended the commit message and PR description, PTAL.

RaduBerinde commented:
LGTM but @andreimatei should also take a look.

Another option would be to disallow `SELECT .. FROM [SHOW TRACE ..]`. One can just use session tracing if they need to select. Not sure if you plan to use this feature in logic tests though,

RaduBerinde commented on pkg/sql/logictest/testdata/logic_test/show_source:
```diff
@@ -221,3 +221,25 @@ CREATE INDEX ix ON foo(x); SELECT * FROM [SHOW TESTING_RANGES FROM INDEX foo@ix]
+# Tracing
+
+statement ok
+BEGIN; SET trace = on; SELECT 1; SET trace = off; COMMIT
```
Is the BEGIN/COMMIT necessary here? Also, can all this be merged with the following query?

knz replied:
Added a comment to clarify.

We can merge but since the following *two* statements use the trace I thought it was clearer to place the set-up state in a distinct line.

RaduBerinde replied:
Cool. Yeah, that's fine, I was just curious if that would work.

knz committed:
sql: generalize the FROM [...] syntax and PREPARE

This patch extends the statement-as-data-source syntax to support any
statement that can also be EXPLAINed. This removes the need for
multiple separate grammar rules, and introduces support for things
like `SELECT ... FROM [INSERT ... RETURNING ...] WHERE ...`,
or more crazy like
`INSERT INTO ... SELECT * FROM [INSERT ... RETURNING ...]`

It also makes all EXPLAINable statements valid for use via PREPARE.

knz commented:
Ok I have extended/simplified the code, PTAL

RaduBerinde commented:
LGTM

knz committed:
sql: simplify SHOW TRACE FOR to use session tracing

knz commented:
@andreimatei the last commit replaces explainTraceNode to use session tracing, but in this PR I am *not* introducing EXPLAIN(DEBUG) functionality yet - will do a separate PR after some experimentation about best way to provide flags.

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> excellent

---

Reviewed 1 of 23 files at r2, 9 of 21 files at r3, 3 of 9 files at r4.
Review status: 11 of 26 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/sql/expand_plan.go, line 91 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16524#-KmgpCBsqORkWUer-nGB:-KmgpCBsqORkWUer-nGC:b-wl7swb) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbe9f8774d3884a856f81bc2d1b838cbc4f8f3e/pkg/sql/expand_plan.go#L91)):*
> ```Go
> 			return plan, err
> 		}
> 		n.plan.MarkDebug(explainDebug)
> ```

I think this is what made INSERT and friends not actually execute. Good riddance!

---

*[pkg/sql/session.go, line 577 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16524#-Kmgp6LeFNbwE6K58Vz3:-Kmgp6LeFNbwE6K58Vz4:bb7r7eq) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbe9f8774d3884a856f81bc2d1b838cbc4f8f3e/pkg/sql/session.go#L577)):*
> ```Go
> // the hijack is no longer needed.
> // TODO(andrei): delete this when explainTraceNode is refactored to use session tracing.
> func (s *Session) hijackCtx(ctx context.Context) func() {
> ```

woot

---

*[pkg/sql/trace.go, line 30 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16524#-KmgmbxrXU7KD3IAywCY:-Kmgmbxs4Cu2bdnzIPQm:b-khc2s7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db459f6a0514bf48208f9c233fcca9def79c4d5/pkg/sql/trace.go#L30)):*
> ```Go
> )
> 
> // traceNode is a planNode that wraps another node and converts DebugValues() results to a
> ```

The comment about DebugValues is now stale?

Also, I think you augment the comment and say how this uses session tracing.

---

*[pkg/sql/trace.go, line 39 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16524#-Kmgmor2UmJumwvRz_xk:-Kmgmor2UmJumwvRz_xl:bcl4y76) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db459f6a0514bf48208f9c233fcca9def79c4d5/pkg/sql/trace.go#L39)):*
> ```Go
> 	execDone bool
> 
> 	traceRows [][7]parser.Datum
> ```

let's now put an alias on `[][7]Datum` (or maybe just `[7]Datum`) since it's getting to be used in a couple of places.

---

*[pkg/sql/trace.go, line 49 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16524#-KmgnNC1jf5P1AZmZm0P:-KmgnNC1jf5P1AZmZm0Q:b-mup6ye) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db459f6a0514bf48208f9c233fcca9def79c4d5/pkg/sql/trace.go#L49)):*
> ```Go
> 
> func (p *planner) makeTraceNode(plan planNode) (planNode, error) {
> 	desc, err := p.getVirtualTabler().getVirtualTableDesc(&sessionTraceTableName)
> ```

"tabler" lol

---

*[pkg/sql/trace_test.go, line 86 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16524#-KmglJzM-nnirSoGmLqX:-KmglJzM-nnirSoGmLqY:b-4v9438) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbe9f8774d3884a856f81bc2d1b838cbc4f8f3e/pkg/sql/trace_test.go#L86)):*
> ```Go
> }
> 
> func TestExplainTrace(t *testing.T) {
> ```

I think this is the time to combine this and the next test.
Perhaps just remove this one and change the other one to use `SHOW TRACE FOR`. (If I remember well, the other one is a slightly more manicured test).

---

*[pkg/sql/parser/explain.go, line 36 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16524#-Kmgg3GNeo69Sa3OIPEm:-Kmgg3GNeo69Sa3OIPEn:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/57222021d9073985754049207542996db0234adc/pkg/sql/parser/explain.go#L36)):*
> ```Go
> 	// Enclosed is set to true if the EXPLAIN syntax was used as a data
> 	// source, and thus enclosed in square brackets.
> 	Enclosed bool
> ```

:thumbsup: 

---

*[pkg/sql/parser/stmt.go, line 430 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16524#-Kmgj5m4fqdqRtb7a5dx:-Kmgj5m4fqdqRtb7a5dy:b-nkdt4u) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbe9f8774d3884a856f81bc2d1b838cbc4f8f3e/pkg/sql/parser/stmt.go#L430)):*
> ```Go
> 
> // StatementTag returns a short string identifying the type of statement.
> func (*ShowTrace) StatementTag() string { return "SHOW DATABASES" }
> ```

wrong tag

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16524#-:-KmgpHd2l2v641AC8PF3:bpxj4br)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 11 of 26 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/sql/trace.go, line 30 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16524#-KmgmbxrXU7KD3IAywCY:-KmhqayGpk4c70MZWY9l:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db459f6a0514bf48208f9c233fcca9def79c4d5/pkg/sql/trace.go#L30)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

The comment about DebugValues is now stale?

Also, I think you augment the comment and say how this uses session tracing.
</blockquote></details>

Done.

---

*[pkg/sql/trace.go, line 39 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/16524#-Kmgmor2UmJumwvRz_xk:-Kmhr3gAeOFgP_38Mm_t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8db459f6a0514bf48208f9c233fcca9def79c4d5/pkg/sql/trace.go#L39)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

let's now put an alias on `[][7]Datum` (or maybe just `[7]Datum`) since it's getting to be used in a couple of places.
</blockquote></details>

Done.

---

*[pkg/sql/trace_test.go, line 86 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16524#-KmglJzM-nnirSoGmLqX:-KmhxNLwkwAvrBHXMRSm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbe9f8774d3884a856f81bc2d1b838cbc4f8f3e/pkg/sql/trace_test.go#L86)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think this is the time to combine this and the next test.
Perhaps just remove this one and change the other one to use `SHOW TRACE FOR`. (If I remember well, the other one is a slightly more manicured test).
</blockquote></details>

Done.

---

*[pkg/sql/parser/stmt.go, line 430 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16524#-Kmgj5m4fqdqRtb7a5dx:-KmhsBE5gC_8DLUF-A6b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bfbe9f8774d3884a856f81bc2d1b838cbc4f8f3e/pkg/sql/parser/stmt.go#L430)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

wrong tag
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16524)*
<!-- Sent from Reviewable.io -->

knz commented:
TFYRs!
